### ISSUE TITLE
Using previous states to suggest parameters

### ISSUE THREAD

**Issue by sachinruk** [optuna/optuna#2311]
Due to the way that our infrastructure is set up I unfortunately cannot attach any kind of Database to our kubernetes cluster. However, I can write out the previous parameters and the result to s3. So what I'm hoping to do is the following workflow (and I'm hoping that there is a way I can use Optuna this way).
1. Generate some initial random parameters (say 5 sets) and run those in parallel using k8s cluster.
2. Record the validation accuracy and write out the set of parameters and its corresponding accuracy to a text file, and save it in s3.
3. Have a node that reads in all past parameters and results and decide what the next 5 sets of parameters are. <- This is where Optuna would hopefully come in.
4. Pass above decided parameters into the next set of parallel pods and run training job as before.

So number 3 above is what I don't know how to do in Optuna. First question is, would Optuna be the right tool here, and if so would appreciate any help/ pointers. If not, could I use a different tool for this process besides using random parameters.

**Comment by crcrpar** [comment]
Hi, I'm not quite confident, but I hope I could provide some notes here.

**Passing five results to Optuna**: `optuna.study.Study.add_trial` enables us to pass the used hyperparameters and the corresponding result to `Study`. [Docs](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.add_trial).

**Suggesting next 5 sets of parameters**: Especially I'm not quite sure but just running optuna.study.Optimize for some trials in your Kubernetes cluster would do what you want.

Hope this helps you a little bit 🙂

**Comment by sachinruk** [comment]
Thanks @crcrpar, certainly helps a big part of the problem that I was having. If I understand the second half of your answer, you're saying simply do `study.optimize` on each pod separately right?

The thing that I'm worried about is that the parameters generated in each of those pods might be the same. I understand that this is a low probability. If we can generate 5 unique set of parameters that I can pass to the pod, I can avoid this situation all together.

**Comment by crcrpar** [comment]
Yes, I meant that and I noticed that parallel execution basically requires some RDB.
Ah, identical sets of parameters could be suggested.

[Ask and Tell API](https://github.com/optuna/optuna/releases/tag/v2.5.0) introduced in v2.5.0 can be useful, I guess.
@hvy What's your take?

**Comment by sachinruk** [comment]
Ah, so I can just:
```python
x = set()
while len(x) < 5:
    trial = study.ask()  # `trial`
    x.add(trial.suggest_float("x", -1, 1))
    
# parse above set of x to pods in parallel <- I can do this part
```
i.e. I'm guessing/ hoping I don't need a corresponding tell component.

**Comment by hvy** [comment]
I think you could use the ask and tell API for that, basically as in the code above. You would then usually tell the results of the corresponding trial/objective values back to the study using `Study.tell`. However, if you instead skip telling and instead write to a bucket or similar, you could create a new study in each pod, each time, parse the data in the bucket, and reconstruct the entire trial history using [`Study.add_trials`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.add_trials). Hope that makes sense.

You will also as you say likely get similar suggestions in your batch. I would suggest avoiding having large batches. It's worth noting that this depends on the sampler (applies to the default `TPESampler` but not `RandomSampler`, etc.) and is actually being worked on as we speak in https://github.com/optuna/optuna/pull/2217. The constant liar algorithm is a variation of the TPE algorithm where parameters near running trials (in Optuna lingo) are less likely to be suggested.

**Comment by sachinruk** [comment]
Just expanding on what you said in the second paragraph, are you saying that the suggestions would be similar because its only conditioning on the past trials, but not the parameters we are trying to run in parallel? So this would dampen our effectiveness of using parallelism right?

Is the PR mentioned above supposed to help with this issue? Or alternately, is there an algorithm that would be better suited besides TPE to tackle this? 

Apologies if I'm making this more complicated than it needs to be.

**Comment by hvy** [comment]
> ... are you saying that the suggestions would be similar because its only conditioning on the past trials, but not the parameters we are trying to run in parallel? So this would dampen our effectiveness of using parallelism right?

Yes and yes.

> Is the PR mentioned above supposed to help with this issue? Or alternately, is there an algorithm that would be better suited besides TPE to tackle this?

For the first question, yes. However, it seems like it won't be ready just yet. There is also a working constant liar sampler as a gist by @toshihikoyanase. Would you be able to use this one perhaps? https://gist.github.com/toshihikoyanase/d2bbf883f4f4acde1793ccc7effee4b7

**Comment by sachinruk** [comment]
Perfect, looks like exactly what I was after. Thank you!
